File: scripts/python/sync_nse_universe.py
#!/usr/bin/env python3
"""
Sync NSE universe from official NSE API to our company_registry table.

Fetch all ~2000 NSE equity symbols + metadata, upsert to Postgres.
Run weekly to catch new listings.
"""

import httpx
import pandas as pd
from datetime import datetime
from sqlalchemy import create_engine, text
import logging
import os

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

DB_URL = os.getenv('DATABASE_URL', 'postgresql://localhost/stockstory')
engine = create_engine(DB_URL)


async def fetch_nse_universe() -> list:
    nse_url = "https://www.nseindia.com/api/equity-stockIndices?index=NIFTY%20500"

    async with httpx.AsyncClient() as client:
        try:
            response = await client.get(
                nse_url,
                headers={"Referer": "https://www.nseindia.com/"}
            )
            data = response.json()

            symbols = []
            for record in data.get('data', []):
                symbols.append({
                    'symbol': record['symbol'],
                    'company_name': record['name'],
                    'sector': record.get('sector', 'General'),
                    'industry': record.get('industry', ''),
                    'market_cap': record.get('marketCap'),
                    'pe': record.get('pe'),
                    'nse_listed': True,
                    'last_updated': datetime.now()
                })

            logger.info(f"Fetched {len(symbols)} NSE symbols")
            return symbols
        except Exception as e:
            logger.error(f"NSE fetch failed: {e}")
            return []


def upsert_company_registry(symbols: list):
    if not symbols:
        return

    df = pd.DataFrame(symbols)

    with engine.connect() as conn:
        for _, row in df.iterrows():
            conn.execute(text("""
                INSERT INTO company_registry (
                    symbol, company_name, sector, industry, 
                    market_cap, pe, nse_listed, active, created_at, updated_at
                ) VALUES (
                    :symbol, :name, :sector, :industry,
                    :market_cap, :pe, :nse_listed, true, now(), now()
                )
                ON CONFLICT (symbol) DO UPDATE SET
                    company_name = EXCLUDED.company_name,
                    sector = EXCLUDED.sector,
                    industry = EXCLUDED.industry,
                    market_cap = EXCLUDED.market_cap,
                    pe = EXCLUDED.pe,
                    updated_at = now()
            """), {
                'symbol': row['symbol'],
                'name': row['company_name'],
                'sector': row['sector'],
                'industry': row['industry'],
                'market_cap': row['market_cap'],
                'pe': row['pe'],
                'nse_listed': True
            })

        conn.commit()
        logger.info(f"Upserted {len(df)} symbols to company_registry")


def validate_chennpetro():
    with engine.connect() as conn:
        registry = conn.execute(text(
            "SELECT symbol, active FROM company_registry WHERE symbol = 'CHENNPETRO'"
        )).scalar()

        if registry:
            logger.info("CHENNPETRO found in registry")
        else:
            logger.warning("CHENNPETRO NOT in registry — will be added by upsert")

        universe = conn.execute(text(
            "SELECT symbol FROM stock_fundamentals WHERE symbol = 'CHENNPETRO' LIMIT 1"
        )).scalar()

        if universe:
            logger.info("CHENNPETRO found in backend universe (fundamentals)")
        else:
            logger.warning("CHENNPETRO NOT in fundamentals — needs EOD sync")


import asyncio


async def main():
    logger.info("NSE Universe Sync Starting...")

    symbols = await fetch_nse_universe()
    upsert_company_registry(symbols)
    validate_chennpetro()

    logger.info("NSE universe sync complete")


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/sync-nse-universe.ts
#!/usr/bin/env node
/**
 * Sync all NSE symbols to company_registry
 * Runs weekly (Sunday 6:30 PM IST)
 * Covers NIFTY 500, MidCap, SmallCap indices
 */

import pg from 'pg';
import axios from 'axios';

const db = new pg.Client({ connectionString: process.env.DATABASE_URL! });

interface Company {
  symbol: string;
  name: string;
  sector: string;
  market_cap?: number;
  listed_date?: string;
  isin?: string;
}

async function fetchNSESymbols(): Promise<Company[]> {
  const companies: Company[] = [];

  try {
    const res = await axios.get(
      'https://www.nseindia.com/content/indices/ind_nifty500list.csv',
      { timeout: 10000 }
    );

    const lines = res.data.split('\n');
    for (let i = 1; i < lines.length; i++) {
      const parts = lines[i].split(',');
      if (parts.length < 2) continue;
      companies.push({
        symbol: parts[0].trim(),
        name: parts[1].trim(),
        sector: parts[3] || 'Unknown',
        isin: parts[2] || undefined,
      });
    }
  } catch (err) {
    console.warn('[Universe] NSE CSV failed, trying Upstox...', (err as Error).message);
    try {
      const res = await axios.get(
        'https://api-v2.upstox.com/market/instruments/?instrument_type=EQUITY',
        {
          timeout: 10000,
          headers: { Authorization: `Bearer ${process.env.UPSTOX_API_KEY!}` },
        }
      );

      for (const item of res.data.data || []) {
        if (item.exchange === 'NSE') {
          companies.push({
            symbol: item.symbol,
            name: item.name,
            sector: item.sector || 'Unknown',
            isin: item.isin,
          });
        }
      }
    } catch (err2) {
      console.error('[Universe] Both sources failed:', err2);
      throw err2;
    }
  }

  return companies;
}

async function upsertRegistry(companies: Company[]): Promise<void> {
  const query = `
    INSERT INTO company_registry (
      symbol, name, sector, market_cap, listed_date, isin
    ) VALUES ($1, $2, $3, $4, $5, $6)
    ON CONFLICT (symbol) DO UPDATE SET
      name = EXCLUDED.name,
      sector = EXCLUDED.sector,
      updated_at = NOW()
  `;

  let count = 0;
  for (const co of companies) {
    try {
      await db.query(query, [
        co.symbol, co.name, co.sector, co.market_cap, co.listed_date, co.isin,
      ]);
      count++;
    } catch (err) {
      console.error(`[Universe] Insert failed for ${co.symbol}:`, err);
    }
  }
  console.log(`[Universe] Upserted ${count}/${companies.length} companies`);
}

async function validateCHENNPETRO(): Promise<boolean> {
  const res = await db.query(
    "SELECT symbol FROM company_registry WHERE symbol = 'CHENNPETRO'"
  );
  return res.rows.length > 0;
}

async function main() {
  await db.connect();
  try {
    console.log('[Universe] Fetching NSE symbols...');
    const companies = await fetchNSESymbols();
    console.log(`[Universe] Fetched ${companies.length} companies`);

    await upsertRegistry(companies);

    const hasChennpetro = await validateCHENNPETRO();
    if (hasChennpetro) {
      console.log('[Universe] CHENNPETRO verified in registry');
    } else {
      console.warn('[Universe] CHENNPETRO not found!');
    }
  } catch (err) {
    console.error('[Universe] Fatal error:', err);
    process.exit(1);
  } finally {
    await db.end();
  }
}

main();
